book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+
+- Method location is at `zebra-chain/src/transaction/sapling/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<AllowNegative> {
+
+}
+```
+
+#### Create a method in `orchard::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/orchard/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<AllowNegative> {
+
+}
+```
+
+#### Create the `Transaction` method
+
+- Method location: `zebra-chain/src/transaction.rs`
+
+```rust
+/// utxos must contain the utxos of every input in the transaction
+/// TODO: what about UTXOs created and spent in the transaction?
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> Result<ValueBalance<NegativeAllowed>, Err> {
+    match self {
+        Transaction::V1 { .. } => {},
+        Transaction::V2 { .. } => {},
+        Transaction::V3 { .. } => {},
+        Transaction::V4 { .. } => {},
+        Transaction::V5 { .. } => {
+            
+        },
+    }
+}
+```
+
+### Create a method in `Block` that returns `ValueBalance<NegativeAllowed>` for the block
+
+- Method location is at `zebra-chain/src/block.rs`.
+- Method will make use of `Transaction::value_balance` method created before.
+
+```rust
+/// utxos must contain the utxos of every input in the block
+/// TODO: what about UTXOs created and spent in the block?
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+    self.transactions()
+        .map(Transaction::value_balance)
+        .sum()
+        .unwrap_or_else(ValueBalance::default)
+}
+```
+
+### Pass the value balance for this block from the consensus into the state
+
+- Add a new field into `PreparedBlock` located at `zebra-state/src/request.rs`, this is the `NonFinalized` section of the state.
+
+```rust
+pub struct PreparedBlock {
+    ..
+    /// The value balances for each pool for this block.
+    pub block_value_balance: ValuePool<NegativeAllowed>,
+}
+```
+- In `zebra-consensus/src/block.rs` pass the value balance to the zebra-state:
+
+```rust
+let block_value_balance = block.value_balance();
+let prepared_block = zs::PreparedBlock {
+    ..
+    block_value_balance,
+};
+```
+
+### Add a value pool into the state `Chain` struct
+
+- This is the value pool for the non finalized part of the blockchain.
+- Location of the `Chain` structure where the pool field will be added: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+pub struct Chain {
+    ..
+    value_pool: ValueBalance<NonNegative>,
+}
+```
+
+### Update `update_chain_state_with()` for `Chain` struct to calculate Chain.value_pool
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+fn update_chain_state_with(&mut self, prepared: &PreparedBlock) {        
+    ...
+    self.value_pool += prepared.block_value_balance;
```

Comment:
Let's `impl UpdateWith<PreparedBlock> for ValueBalance` separately, so we can document how it checks the consensus rules.

Reply:
https://github.com/ZcashFoundation/zebra/pull/2430/commits/b9adea207d9c063ee92e9afbb178155b3c412ce6